### ISSUE TITLE
BindingContext for type resolution doesn't take friend relations into account

### ISSUE THREAD

**Issue by TWiStErRob** [detekt/detekt#5396]
## Expected Behavior
No errors or warnings in Detekt, code compiles and executes just fine.

## Observed Behavior
```
src\test\kotlin\net\twisterrob\cinema\cineworld\quickbook\QuickbookErrorsKtTest.kt:16:3: error: 'public open val errors: List<String>? defined in net.twisterrob.cinema.cineworld.quickbook.QuickbookErrorsKtTest.FakeResponse' has no access to 'public abstract val errors: List<String>? defined in net.twisterrob.cinema.cineworld.quickbook.QuickbookErrors', so it cannot override it
                override val errors: List<String>?
                ^
src\test\kotlin\net\twisterrob\cinema\cineworld\quickbook\QuickbookErrorsKtTest.kt:17:6: error: cannot access 'QuickbookErrors': it is internal in 'net.twisterrob.cinema.cineworld.quickbook'
        ) : QuickbookErrors
            ^
src\test\kotlin\net\twisterrob\cinema\cineworld\quickbook\QuickbookErrorsKtTest.kt:22:33: error: cannot access 'throwErrorOrReturn': it is internal in 'net.twisterrob.cinema.cineworld.quickbook'
                val result: String = response.throwErrorOrReturn { it.data }
                                              ^
src\test\kotlin\net\twisterrob\cinema\cineworld\quickbook\QuickbookErrorsKtTest.kt:30:33: error: cannot access 'throwErrorOrReturn': it is internal in 'net.twisterrob.cinema.cineworld.quickbook'
                val result: String = response.throwErrorOrReturn { it.data }
                                              ^
src\test\kotlin\net\twisterrob\cinema\cineworld\quickbook\QuickbookErrorsKtTest.kt:38:53: error: cannot access 'throwErrorOrReturn': it is internal in 'net.twisterrob.cinema.cineworld.quickbook'
                assertThrows<QuickbookErrorsException> { response.throwErrorOrReturn { it.data } }.let { result ->
                                                                  ^
src\test\kotlin\net\twisterrob\cinema\cineworld\quickbook\QuickbookErrorsKtTest.kt:50:53: error: cannot access 'throwErrorOrReturn': it is internal in 'net.twisterrob.cinema.cineworld.quickbook'
                assertThrows<QuickbookErrorsException> { response.throwErrorOrReturn { it.data } }.let { result ->
                                                                  ^
The BindingContext was created with 6 issues. Run detekt with --debug to see the error messages.
```

## Steps to Reproduce
 * https://github.com/TWiStErRob/repros/tree/master/detekt/BindingContext-internal-test
 * `gradlew cleanDetektTest detektTest`

## Context
`src/test/kotlin` has a friend access to `src/main/kotlin` so that `internal` program elements can be tested. It seems Detekt doesn't know this.

## Your Environment

* Version of detekt used: 1.21.0
* Version of Gradle used (if applicable): 7.5.1
* Link to your project (if it's a public repository): repro was extracted from https://github.com/TWiStErRob/net.twisterrob.cinema/blob/a07c718574153d80da3647ffdc1ef97ab3e4c98c/Heroku/backend/quickbook/src/test/kotlin/net/twisterrob/cinema/cineworld/quickbook/QuickbookErrorsKtTest.kt#L16

**Comment by TWiStErRob** [comment]
Note: the above is a trivial example, in another module it has a lot of knock-on effects, making Detekt rule execution weaker I guess?
![image](https://user-images.githubusercontent.com/2906988/194749363-4243a45c-540c-4398-aa5e-4752ac85f7c7.png)

**Comment by TWiStErRob** [comment]
Additionally (probably the same issue but with `private` within the same file):
```
> Task :backend:database:detektTest
Heroku\backend\database\src\test\kotlin\net\twisterrob\cinema\database\model\test\ModelIntgTestExtension.kt:52:5: error: cannot access 'ModelIntgTestExtensionComponent': it is private in file
                        .graphDBUri(testServer.boltURI())
    ^
Heroku\backend\database\src\test\kotlin\net\twisterrob\cinema\database\model\test\ModelIntgTestExtension.kt:53:5: error: cannot access 'ModelIntgTestExtensionComponent': it is private in file
                        .build()
    ^
The BindingContext was created with 2 issues. Run detekt with --debug to see the error messages.
```

which is pointing to this code in `src\test\kotlin\net\twisterrob\cinema\database\model\test\ModelIntgTestExtension.kt`:
```kotlin
class ModelIntgTestExtension : BeforeEachCallback {
	override fun beforeEach(extensionContext: ExtensionContext) {
		val dagger = DaggerModelIntgTestExtensionComponent
			.builder()
			.graphDBUri(testServer.boltURI())
			.build()
		extensionContext.store.put(dagger)
	}
}

@Component(modules = [Neo4JModule::class])
private interface ModelIntgTestExtensionComponent {

	val session: Session

	@Component.Builder
	interface Builder {
		@BindsInstance
		fun graphDBUri(@Named(GRAPH_DB) uri: URI): Builder
		fun build(): ModelIntgTestExtensionComponent
	}
}
```
and generated code in `build\generated\source\kapt\test\net\twisterrob\cinema\database\model\test\DaggerModelIntgTestExtensionComponent.java`:
```java
final class DaggerModelIntgTestExtensionComponent {
  private DaggerModelIntgTestExtensionComponent() {
  }

  public static ModelIntgTestExtensionComponent.Builder builder() {
    return new Builder();
  }
```

**Comment by 3flex** [comment]
Thanks for the report, this is a known issue: #3067
